src/pudl/transform/classes.py

```diff
@@ -156,7 +163,21 @@ def multicol_transform_factory(
     col_func: ColumnTransformFunc,
     drop=True,
 ) -> MultiColumnTransformFunc:
-    """A factory for creating a multi-column transform function."""
+    """A factory function for creating multi-column transform functions.
+
+    Converts individual column transformers into multi-column transformers. This saves
+    us from having to manually iterate over dataframes in many places, applying the
+    same transform functions with different parameters to multiple columns. Instead we
+    can define a function that transforms a column given some parameters, and then
+    easily apply that function to many columns using a dictionary of parameters. Uniform
+    logging output is also integrated into the multi-column transform function.
+
+    Args:
+        col_func: A single column transform function.
```

Comment:
**All** multicolumn params are just dictionaries of the single column params associated with the single column transform function. The keys in that dictionary are the names of the column that each parameter applies to. The factory function doesn't know what type of params it's going to be dealing with. This will be easy to state explicitly when Pydantic 2.0 comes out, but it's handwavy in the `MultiColumnTransformParams` right now.

What kind of documentation would help here? I'm not sure what kind of example you're thinking of.  It's just a dictionary of `TransformParams`. We could point folks to the big dictionaries in the params module?

Reply:
I'm suggesting adding an example like the one I was working out above. this `multicol_transform_factory` requires layers of setup to use properly. i've re-giggered the example I was posing above and cannot get it to work.

```
Ex:
class AddNumParam(TransformParams):
     num_value: int

def add_num_col(col: pd.Series, param: AddNumParam):
    return col + param.num_value

add_num_col_multicol = multicol_transform_factory(add_num_col)

class AddNumMultiCol(MultiColumnTransformParams):
    params: dict[str, AddNumParam]
    # maybe?
    
add_num_multicol_params = AddNumMultiCol(
    params={
        "col1": AddNumParam(num_value=1),
        "col2": AddNumParam(num_value=2)
    }
)

df = pd.DataFrame(
    {
        "col1": [1, 2, 3, 4],
        "col2": [10, 12, 14, 16]
    }
)

df_new = add_num_col_multicol(df, add_num_multicol_params)

# the expected outcome
df_out = pd.DataFrame(
    {
        "col1": [i + 1 for i in [1, 2, 3, 4]],
        "col2": [i + 2 for i in [10, 12, 14, 16]]
    }
)
pd.testing.assert_frame_equal(df_new, df_out)
```

The above code fails w/ the error below presumably bc I do not know how to setup `AddNumMultiCol` properly

```
File ~/code/pudl/src/pudl/transform/classes.py:172, in multicol_transform_factory.<locals>.InnerMultiColumnTransformFunc.__call__(self, df, params)
    170 for col_name in params:
    171     logger.info(col_name)
--> 172     if col_name in df.columns:
    173         logger.debug(f"Applying {col_func.__name__} to {col_name}")
    174         new_col = col_func(col=df[col_name], params=params[col_name])

File ~/mambaforge/envs/pudl-dev/lib/python3.10/site-packages/pandas/core/indexes/base.py:5022, in Index.__contains__(self, key)
   4987 def __contains__(self, key: Any) -> bool:
   4988     """
   4989     Return a boolean indicating whether the provided key is in the index.
   4990 
   (...)
   5020     False
   5021     """
-> 5022     hash(key)
   5023     try:
   5024         return key in self._engine

TypeError: unhashable type: 'dict'
```

Reply:
even though folks should probably be using the big params dictionaries, it is my understanding that these elements are able to be compiled and used without needing a whole big params dictionary. i'm suggesting here an example that breaks down how these pieces fit together in a more decomposed state.

Reply:
okay after much trying and looking through the unit tests i think I am doing this correctly

```
# setup
class AddNumParam(TransformParams):
     num_value: int

def add_num_col(col: pd.Series, params: AddNumParam):
    return col + params.num_value

add_num_col_multicol = multicol_transform_factory(add_num_col)

# use
df = pd.DataFrame(
    {
        "col1": [1, 2, 3, 4],
        "col2": [10, 12, 14, 16]
    }
)

df_new = add_num_col_multicol(
    df,
    params={
        "col1": AddNumParam(num_value=1),
        "col2": AddNumParam(num_value=2)
    }
)

df_out = pd.DataFrame(
    {
        "col1": [i + 1 for i in [1, 2, 3, 4]],
        "col2": [i + 2 for i in [10, 12, 14, 16]]
    }
)

pd.testing.assert_frame_equal(df_new, df_out)
```

Reply:
maybe with this chain of examples of how i was working through this you can see where my gaps in understanding lied. but also very happy to chat about it more broadly

Reply:
I adapted this example slightly and integrated it into the docstring.